crowdsourcehinter/static/js/src/crowdsourcehinter.js

```diff
@@ -0,0 +1,305 @@
+
+    /**
+     * Get a hint to show to the student after incorrectly answering a question.
+     * @param data is data generated by the problem_graded event
+     */
+    function get_hint(data){
+        $(".crowdsourcehinter_block", element).show();
+        $.ajax({
+            type: "POST",
+            url: runtime.handlerUrl(element, 'get_hint'),
+            data: JSON.stringify({"submittedanswer": unescape(data[0])}),
+            success: showHint
+        });
+    }
+
+    /**
+     * Start student hint feedback. This function is called after the student answers
+     * the question correctly.
+     */
+    function start_feedback(){
+        $('.csh_correct', element).show();
+        $(".csh_hint_reveal", element).hide();
+        //send empty data for ajax call because not having a data field causes error
+        $.ajax({
+            type: "POST",
+            url: runtime.handlerUrl(element, 'get_feedback'),
+            data: JSON.stringify({}),
+            success: showStudentContribution
+        });
+    }
+
+    /**
+     * Check whether student answered the question correctly and call the appropriate
+     * function afterwards. Current method for determining correctness if very brittle.
+     * @param event_type, element are both unused but automatically passed
+     * @param data is generated by problem_graded event, contains status and data of the problem block
+     */
+    function onStudentSubmission(){ return function(event_type, data, element){
+        //search method of correctness of problem is brittle due to checking for a class within
+        //the problem block.
+        if (data[1].search(/class="correct/) === -1){
+            get_hint(data);
+        } else { //if the correct answer is submitted
+            start_feedback();
+        }
+    }}
+    Logger.listen('problem_graded', data.hinting_element, onStudentSubmission());
+
+    /**
+     * Modify csh_Hints attributes to show hint to the student.
+     */
+    function showHint(result){
+        $('.csh_Hints', element).attr('student_answer', result.StudentAnswer);
+        $('.csh_Hints', element).attr('hint_received', result.Hints);
+        $('.csh_Hints', element).text("Hint: " + result.Hints);
+        Logger.log('crowd_hinter.showHint', {"student_answer": result.StudentAnswer, "hint_received": result.Hints});
+    }
+
+    /**
+     * Called by showStudentContribution to append hints into divs created by
+     * showStudentSubmissoinHistory, after the student answered the question correctly.
+     * Feedback on hints at this stage consists of upvote/downvote/report buttons.
+     * @param hint is the first hint that was shown to the student
+     * @param student_answer is the first incorrect answer submitted by the student
+     */
+    function showHintFeedback(hint, student_answer){
+        $(".csh_student_answer", element).each(function(){
+            if ($(this).find('.csh_answer_text').attr('answer') == student_answer){
```

Comment:
c. Kill if statement. Regardless, it's better to do this with a jquery selector than iterate/if.